### PR TITLE
blockchain: Add ReconsiderBlock()

### PR DESCRIPTION
**Pull request by kcalvinalvin**
Adds `ReconsiderBlock()` on blockchain that will clear up any block validation statuses and re-validate the block. If the block is part of a longer chain, then the chain will reorganize to that chain.

Depends on #2155

### CODE REVIEW

**Comment by BrannonKing** [comment]
I just want to note that the InvalidateBlock/ReconsiderBlock in lbcd and and bchd is incorrect. See https://github.com/gcash/bchd/issues/512

**Comment by Roasbeef** [comment]
Can be rebased now!

**Comment by kcalvinalvin** [comment]
> I just want to note that the InvalidateBlock/ReconsiderBlock in lbcd and and bchd is incorrect. See [gcash/bchd#512](https://github.com/gcash/bchd/issues/512)

This is a completely independent implementation by me. I didn't even know they were implemented there.

File: blockchain/chain.go [hunk 1]
```diff
@@ -1949,6 +1949,88 @@ func (b *BlockChain) InvalidateBlock(hash *chainhash.Hash) error {
...
+		}
+
+		// Mark the current tip as the tip being reconsidered.
+		reconsiderTip = tip
+
+		// Unset the status of all the parents up until it reaches the block
+		// being reconsidered.
+		for n := tip; n != nil && n != node; n = n.parent {
+			b.index.UnsetStatusFlags(n, statusInvalidAncestor)
+		}
```
Comment (Roasbeef) [inline blockchain/chain.go:2001]: Can't we break after this?

Also isn't it possible that the block we want to reconsider is a descendent of _several_ of the inactive chain tips? If this is the case, then we'll continually re-assign `reconsiderTip` for each of them. 

Consider a scenario where we had 3 inactive chain tips `{A, B, C}`, block `D` that we want to reconsider is a descendent of all 3 chain tips, but they later on fork further down the line produce distinct chain tips (that we thought were invalid in the past).

File: blockchain/chain.go [hunk 2]
```diff
@@ -1949,6 +1949,88 @@ func (b *BlockChain) InvalidateBlock(hash *chainhash.Hash) error {
...
+	// the blockindex after we call this function.
+	_, _, _, err := b.verifyReorganizationValidity(detachNodes, attachNodes)
+	if writeErr := b.index.flushToDB(); writeErr != nil {
+		log.Warnf("Error flushing block index changes to disk: %v", writeErr)
+	}
+	if err != nil {
+		// If we errored out during the verification of the reorg branch,
+		// it's ok to return nil as we reconsidered the block and determined
+		// that it's invalid.
+		return nil
```
Comment (Roasbeef) [inline blockchain/chain.go:2028]: Should log the error here.

File: blockchain/chain.go (hunk 1 above)
Comment (Roasbeef) [reply blockchain/chain.go:2001]: Actually on second thought, block hash chains are unique, so if `D` is a descendent of all 3 chain tips, then the chain tips should themselves be identical.

File: blockchain/chain.go [hunk 3]
```diff
@@ -1949,6 +1949,88 @@ func (b *BlockChain) InvalidateBlock(hash *chainhash.Hash) error {
 	return err
 }
 
+// ReconsiderBlock reconsiders the validity of the block with the given hash.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) ReconsiderBlock(hash *chainhash.Hash) error {
+	b.chainLock.Lock()
+	defer b.chainLock.Unlock()
+
```
Comment (Roasbeef) [inline blockchain/chain.go:1958]: `log.Infof("Reconsidering block_hash=%x", hash[:])`

File: blockchain/chain.go [hunk 4]
```diff
@@ -1949,6 +1949,88 @@ func (b *BlockChain) InvalidateBlock(hash *chainhash.Hash) error {
...
+	}
+
+	// Clear the status of the block being reconsidered.
+	b.index.UnsetStatusFlags(node, statusInvalidAncestor)
+	b.index.UnsetStatusFlags(node, statusValidateFailed)
+
+	// Grab all the tips.
+	tips := b.index.InactiveTips(b.bestChain)
+	tips = append(tips, b.bestChain.Tip())
+
```
Comment (Roasbeef) [inline blockchain/chain.go:1978]: `log.Debugf("Examining %v inactive chain tips for reconsideration")`

File: blockchain/chain.go [hunk 5]
```diff
@@ -1949,6 +1949,88 @@ func (b *BlockChain) InvalidateBlock(hash *chainhash.Hash) error {
...
+	// Compare the cumulative work for the branch being reconsidered.
+	if reconsiderTip.workSum.Cmp(b.bestChain.Tip().workSum) <= 0 {
+		return nil
+	}
+
+	// If the reconsider tip has a higher cumulative work, then reorganize
+	// to it after checking the validity of the nodes.
+	detachNodes, attachNodes := b.getReorganizeNodes(reconsiderTip)
+
+	// We're checking if the reorganization that'll happen is actually valid.
```
Comment (Roasbeef) [inline blockchain/chain.go:2013]:
```
log.Infof("Re-org to reconsider_block=%x size %v (num_add=%v, num_detach=%v)", reconsiderTip.hash[;], len(detchNodes)+len(attachNodes), len(detchNodes), len(attachNodes))
```

**Review by Roasbeef — CHANGES_REQUESTED** [review]

**Comment by kcalvinalvin** [comment]
Closing in favor of #2196

**Pull request closed without merging**